File: ts/src/session.ts
import { randomUUID } from "node:crypto";
import type { Advice, Reward, Trace, Adapter } from "./types.js";
import type { Advisor } from "./advisor.js";
import type { RewardScorer } from "./reward.js";

export interface SessionConfig {
  mode: "one-shot" | "multi-turn";
  maxSteps: number;
  rewardThreshold: number;
}

/** Session lifecycle state (mirrors spec/session-protocol.md state machine). */
export enum SessionState {
  Init = "init",
  Advise = "advise",
  Frontier = "frontier",
  Score = "score",
  Trace = "trace",
  Done = "done",
}

/** Result of a single advise->frontier->score->trace step. */
export interface SessionResult {
  output: string;
  reward: Reward;
  trace: Trace;
  state: SessionState;
}

/** Valid state transitions per session-protocol.md. */
const VALID_TRANSITIONS: Record<string, Set<string>> = {
  [SessionState.Init]: new Set([SessionState.Advise, SessionState.Done]),
  [SessionState.Advise]: new Set([SessionState.Frontier]),
  [SessionState.Frontier]: new Set([SessionState.Score]),
  [SessionState.Score]: new Set([SessionState.Trace]),
  [SessionState.Trace]: new Set([SessionState.Advise, SessionState.Done]),
};

export class Session {
  private config: SessionConfig;
  private state: SessionState = SessionState.Init;
  private step = 0;
  private traces: Trace[] = [];

  constructor(
    private advisor: Advisor,
    private adapter: Adapter,
    private scorer: RewardScorer,
    config?: Partial<SessionConfig>,
  ) {
    this.config = {
      mode: config?.mode ?? "one-shot",
      maxSteps: config?.maxSteps ?? 10,
      rewardThreshold: config?.rewardThreshold ?? 1.0,
    };
  }

  /** Current state machine state. */
  getState(): SessionState {
    return this.state;
  }

  /** All traces recorded so far. */
  getTraces(): Trace[] {
    return this.traces;
  }

  /** Current step count. */
  getStep(): number {
    return this.step;
  }

  private transition(to: SessionState): void {
    const allowed = VALID_TRANSITIONS[this.state];
    if (!allowed || !allowed.has(to)) {
      throw new Error(
        `invalid state transition: ${this.state} -> ${to}`,
      );
    }
    this.state = to;
  }

  /** Run a single advise->frontier->score->trace step with a given sessionId. */
  private async runStep(
    sessionId: string,
    prompt: string,
    context: Record<string, unknown>,
  ): Promise<SessionResult> {
    // First step: Init -> Advise. Subsequent steps: Trace -> Advise.
    this.transition(SessionState.Advise);

    const input = { prompt, context };

    let advice: Advice;
    try {
      advice = await this.advisor.generateAdvice(input);
    } catch {
      advice = {
        domain: "unknown",
        steering_text: "",
        confidence: 0,
        version: "1.0",
        constraints: [],
        metadata: {},
      };
    }

    this.transition(SessionState.Frontier);

    const { output, meta: frontierMeta } = await this.adapter.call(
      prompt,
      advice,
    );

    this.transition(SessionState.Score);

    let reward: Reward;
    try {
      reward = await this.scorer.score(output, context);
    } catch {
      reward = { score: NaN, breakdown: {} };
    }

    this.transition(SessionState.Trace);
    this.step += 1;

    const trace: Trace = {
      id: randomUUID(),
      session_id: sessionId,
      timestamp: new Date().toISOString(),
      input,
      advice,
      frontier: frontierMeta,
      reward,
    };

    this.traces.push(trace);

    return { output, reward, trace, state: this.state };
  }

  /**
   * Run a session. Behavior depends on config.mode:
   * - "one-shot": single step, returns one result.
   * - "multi-turn": loops up to maxSteps, stops early if
   *   reward.score >= rewardThreshold. Returns all step results.
   */
  async run(
    prompt: string,
    context: Record<string, unknown> = {},
  ): Promise<SessionResult | SessionResult[]> {
    const sessionId = randomUUID();
    this.step = 0;
    this.traces = [];
    this.state = SessionState.Init;

    if (this.config.mode === "multi-turn") {
      return this.runMultiTurn(sessionId, prompt, context);
    }

    return this.runOneShot(sessionId, prompt, context);
  }

  private async runOneShot(
    sessionId: string,
    prompt: string,
    context: Record<string, unknown>,
  ): Promise<SessionResult> {
    const result = await this.runStep(sessionId, prompt, context);
    this.transition(SessionState.Done);
    return { ...result, state: this.state };
  }

  private async runMultiTurn(
    sessionId: string,
    prompt: string,
    context: Record<string, unknown>,
  ): Promise<SessionResult[]> {
    // Short-circuit: 0-step session transitions Init -> Done.
    if (this.config.maxSteps === 0) {
      this.transition(SessionState.Done);
      return [];
    }

    const results: SessionResult[] = [];

    while (this.step < this.config.maxSteps) {
      const result = await this.runStep(sessionId, prompt, context);
      results.push(result);

      if (result.reward.score >= this.config.rewardThreshold) {
        break;
      }
    }

    this.transition(SessionState.Done);
    return results;
  }
}

File: rs/src/session.rs
use crate::adapter::Adapter;
use crate::advisor::{Advice, Advisor};
use crate::error::FitError;
use crate::reward::{Reward, RewardScorer};
use crate::trace::{Trace, TraceWriter};
use std::collections::HashMap;

/// Session lifecycle state.
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum SessionState {
    Init,
    Advise,
    Frontier,
    Score,
    Trace,
    Done,
}

impl std::fmt::Display for SessionState {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            SessionState::Init => write!(f, "init"),
            SessionState::Advise => write!(f, "advise"),
            SessionState::Frontier => write!(f, "frontier"),
            SessionState::Score => write!(f, "score"),
            SessionState::Trace => write!(f, "trace"),
            SessionState::Done => write!(f, "done"),
        }
    }
}

/// Session configuration.
#[derive(Debug, Clone)]
pub struct SessionConfig {
    pub mode: SessionMode,
    pub max_steps: u32,
    pub reward_threshold: f64,
}

/// Session mode: one-shot or multi-turn.
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum SessionMode {
    OneShot,
    MultiTurn,
}

impl Default for SessionConfig {
    fn default() -> Self {
        Self {
            mode: SessionMode::OneShot,
            max_steps: 10,
            reward_threshold: 1.0,
        }
    }
}

/// Session result.
pub struct SessionResult {
    pub output: String,
    pub reward: Reward,
    pub trace: Trace,
    pub state: SessionState,
}

/// Session orchestrator with full state machine.
///
/// States: Init -> Advise -> Frontier -> Score -> Trace -> (Advise | Done)
pub struct Session<A, S, R>
where
    A: Advisor,
    S: Adapter,
    R: RewardScorer,
{
    advisor: A,
    adapter: S,
    scorer: R,
    config: SessionConfig,
    state: SessionState,
    step: u32,
    traces: Vec<Trace>,
    writer: Option<TraceWriter>,
}

impl<A, S, R> Session<A, S, R>
where
    A: Advisor,
    S: Adapter,
    R: RewardScorer,
{
    /// Create a new session with default config.
    pub fn new(advisor: A, adapter: S, scorer: R) -> Self {
        Self {
            advisor,
            adapter,
            scorer,
            config: SessionConfig::default(),
            state: SessionState::Init,
            step: 0,
            traces: vec![],
            writer: None,
        }
    }

    /// Set session configuration.
    pub fn with_config(mut self, config: SessionConfig) -> Self {
        self.config = config;
        self
    }

    /// Set trace writer for persistent output.
    pub fn with_writer(mut self, writer: TraceWriter) -> Self {
        self.writer = Some(writer);
        self
    }

    /// Get current state.
    pub fn state(&self) -> &SessionState {
        &self.state
    }

    /// Get all recorded traces.
    pub fn traces(&self) -> &[Trace] {
        &self.traces
    }

    /// Validate a state transition.
    fn validate_transition(&self, to: &SessionState) -> Result<(), FitError> {
        let valid = match (&self.state, to) {
            (SessionState::Init, SessionState::Advise) => true,
            (SessionState::Advise, SessionState::Frontier) => true,
            (SessionState::Frontier, SessionState::Score) => true,
            (SessionState::Score, SessionState::Trace) => true,
            (SessionState::Trace, SessionState::Advise) => true,
            (SessionState::Trace, SessionState::Done) => true,
            _ => false,
        };

        if !valid {
            return Err(FitError::InvalidTransition {
                from: self.state.to_string(),
                to: to.to_string(),
            });
        }

        Ok(())
    }

    /// Transition to a new state.
    fn transition(&mut self, to: SessionState) -> Result<(), FitError> {
        self.validate_transition(&to)?;
        self.state = to;
        Ok(())
    }

    /// Run a one-shot session: Init -> Advise -> Frontier -> Score -> Trace -> Done
    pub async fn run(
        &mut self,
        prompt: &str,
        context: HashMap<String, serde_yaml::Value>,
    ) -> Result<SessionResult, FitError> {
        let session_id = uuid::Uuid::new_v4().to_string();

        // Init -> Advise
        self.transition(SessionState::Advise)?;

        // Build input map
        let mut input = HashMap::new();
        input.insert(
            "prompt".to_string(),
            serde_yaml::Value::String(prompt.to_string()),
        );
        for (k, v) in &context {
            input.insert(k.clone(), v.clone());
        }

        // Advise: generate advice, fallback to empty on failure
        let advice = match self.advisor.generate_advice(input.clone()).await {
            Ok(a) => a,
            Err(_) => Advice::new("unknown", "", 0.0),
        };

        // Advise -> Frontier
        self.transition(SessionState::Frontier)?;

        // Frontier: call adapter
        let (output, frontier_meta) = self.adapter.call(prompt, &advice).await?;

        // Frontier -> Score
        self.transition(SessionState::Score)?;

        // Score: evaluate output, fallback to 0 on failure
        let reward = match self.scorer.score(&output, &context) {
            Ok(r) => r,
            Err(_) => Reward::new(0.0, HashMap::new()),
        };

        // Score -> Trace
        self.transition(SessionState::Trace)?;
        self.step += 1;

        let trace = Trace {
            id: uuid::Uuid::new_v4().to_string(),
            session_id: session_id.clone(),
            timestamp: chrono::Utc::now().to_rfc3339(),
            input,
            advice,
            frontier: frontier_meta,
            reward: reward.clone(),
            metadata: {
                let mut m = HashMap::new();
                m.insert(
                    "trace_version".into(),
                    serde_yaml::Value::String("1.0".into()),
                );
                m
            },
        };

        // Write trace if writer is configured
        if let Some(ref writer) = self.writer {
            writer.write(&trace, self.step)?;
        }

        self.traces.push(trace.clone());

        // Trace -> Done (one-shot always completes after first cycle)
        self.transition(SessionState::Done)?;

        Ok(SessionResult {
            output,
            reward,
            trace,
            state: self.state.clone(),
        })
    }

    /// Run a multi-turn session with max_steps limit.
    pub async fn run_multi_turn(
        &mut self,
        prompt: &str,
        mut context: HashMap<String, serde_yaml::Value>,
    ) -> Result<Vec<SessionResult>, FitError> {
        self.config.mode = SessionMode::MultiTurn;
        let mut results = vec![];
        let mut current_prompt = prompt.to_string();

        loop {
            let result = self.run(&current_prompt, context.clone()).await?;
            let done = result.reward.score >= self.config.reward_threshold
                || self.step >= self.config.max_steps;

            // Collect observations for next turn
            context.insert(
                "step".into(),
                serde_yaml::Value::Number(self.step.into()),
            );

            results.push(result);

            if done {
                break;
            }
        }

        Ok(results)
    }
}
